### PR TITLE
Optimize computeElemValues for constant monomials

### PR DESCRIPTION
**Pull request by lindsayad**
- Create `MooseVariableConstMonomial` that overrides `computeElemValues` and makes use of fact that there's only one DOF per element and the elemental value is the same at every quadrature point

Performance summary (with iprofiler) for following input file:

- Old code:
    - Total run time: 7.745 s
    - Time in `MooseVariable::computeElemValues`: 1.256 s
    - % of calculation: 16

- New code:
    - Total run time: 5.766 s
    - Time in `MooseVariableConstMonomial::computeElemValues`: 0.624 s
    - % of calculation : 11

So looks about like a 2x speed-up.

```
[Mesh]
  type = GeneratedMesh
  dim = 2
  nx = 10
  ny = 10
[]

[Variables]
  [./u]
    family = MONOMIAL
    order = CONSTANT
  [../]
[]

[Kernels]
  [./rxn]
    type = Reaction
    variable = u
  [../]
  [./ffn]
    type = UserForcingFunction
    variable = u
    function = 1
  [../]
[]

[Executioner]
  type = Steady

  # Preconditioned JFNK (default)
  solve_type = 'JFNK'
[]

[LotsOfAuxVariables]
  [./avar]
    number = 20000
    family = MONOMIAL
    order = CONSTANT
  [../]
[]
```

Targets #9836

### CODE REVIEW

**Comment by lindsayad** [comment]
There's still some clean-up I can do in `computeElemValues`...

**Comment by andrsd** [comment]
Also benchmark a 3D problem: more q-points per element, you should see a bigger gain.

**Comment by lindsayad** [comment]
Ok for 3D test problem:

- Old code: 
    - Total time: 58.985 s
    - time in `MooseVariable::computeElemValues`: 16.406 s
    - % = 27

- New code:
    - Total time: 38.674 s
    - time in `MooseVariableConstMonomial::computeElemValues`:  5.474 s
    - % = 14

File: framework/src/base/SystemBase.C [hunk 1]
```diff
@@ -506,8 +507,11 @@ SystemBase::addVariable(const std::string & var_name,
   {
     // FIXME: we cannot refer fetype in libMesh at this point, so we will just make a copy in
     // MooseVariableBase.
-    MooseVariable * var =
-        new MooseVariable(var_num, type, *this, _subproblem.assembly(tid), _var_kind);
+    auto * var =
+        (type == (FEType(0, MONOMIAL)))
+            ? new MooseVariableConstMonomial(
+                  var_num, type, *this, _subproblem.assembly(tid), _var_kind)
+            : new MooseVariable(var_num, type, *this, _subproblem.assembly(tid), _var_kind);
```
Comment (permcody) [inline framework/src/base/SystemBase.C]: Holy train-wreck! You should probably split this up into a few lines. I know this was already a raw memory allocation, when you find these, see if they can be converted to smart pointers.

**Comment by lindsayad** [comment]
Ready for further review.

**Review by permcody — APPROVED** [review]
Looks good, I'm not super crazy about add virtuals to MooseVariable, but in this case if it helps with speed, it's probably worth it.

**Comment by permcody** [comment]
Do you plan on doing more on #9836? It wasn't closed by this PR but it looks like it maybe should have been.

**Comment by lindsayad** [comment]
Yea, it should be closed.

**Pull request merged**